opentracing-mock/src/main/java/io/opentracing/mock/MockTracer.java

```diff
@@ -112,6 +119,82 @@ protected void onSpanFinished(MockSpan mockSpan) {
+                if (carrier instanceof Binary) {
+                    Binary binary = (Binary) carrier;
+                    ByteArrayOutputStream stream = new ByteArrayOutputStream();
+                    ObjectOutputStream objStream = null;
+                    try {
+                        objStream = new ObjectOutputStream(stream);
+                        objStream.writeLong(ctx.spanId());
+                        objStream.writeLong(ctx.traceId());
+
+                        for (Map.Entry<String, String> entry : ctx.baggageItems()) {
+                            objStream.writeUTF(entry.getKey());
+                            objStream.writeUTF(entry.getValue());
+                        }
+                        objStream.flush(); // *need* to flush ObjectOutputStream.
+
+                        byte[] buff = stream.toByteArray();
+                        binary.setInjectBufferLength(buff.length);
+                        binary.injectBuffer().put(buff);
+
+                    } catch (IOException e) {
+                        throw new RuntimeException("Corrupted state");
+                    } finally {
+                        if (objStream != null) {
+                            try { objStream.close(); } catch (Exception e2) {}
+                        }
+                    }
+                } else {
+                    throw new IllegalArgumentException("Unknown carrier");
+                }
+            }
+
+            @Override
+            public <C> MockSpan.MockContext extract(Format<C> format, C carrier) {
+                Long traceId = null;
+                Long spanId = null;
+                Map<String, String> baggage = new HashMap<>();
+
+                if (carrier instanceof Binary) {
+                    Binary binary = (Binary) carrier;
+                    ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+                    ObjectInputStream objStream = null;
+                    try {
+                        ByteBuffer extractBuff = binary.extractBuffer();
+                        byte[] buff = new byte[extractBuff.remaining()];
+                        extractBuff.get(buff);
+
+                        objStream = new ObjectInputStream(new ByteArrayInputStream(buff));
+                        spanId = objStream.readLong();
+                        traceId = objStream.readLong();
+
+                        while (objStream.available() > 0) {
+                            baggage.put(objStream.readUTF(), objStream.readUTF());
+                        }
+                    } catch (IOException e) {
+                    } finally {
+                        if (objStream != null) {
+                            try { objStream.close(); } catch (Exception e2) {}
+                        }
+                    }
+                } else {
+                    throw new IllegalArgumentException("Unknown carrier");
```

Comment:
I would move this to the top to reduce nesting. Also clarify `"expecting Binary, received " + carrier.getClass()`

Reply:
Agreed ;)